src/function/scalar/variant/variant_typeof.cpp

```diff
@@ -0,0 +1,108 @@
+#include "duckdb/function/scalar/variant_utils.hpp"
+#include "duckdb/function/scalar/variant_functions.hpp"
+#include "duckdb/common/serializer/varint.hpp"
+#include "duckdb/common/enum_util.hpp"
+
+namespace duckdb {
+
+static void VariantTypeofFunction(DataChunk &input, ExpressionState &state, Vector &result) {
+	auto count = input.size();
+
+	D_ASSERT(input.ColumnCount() == 1);
+	auto &variant = input.data[0];
+	D_ASSERT(variant.GetType() == LogicalType::VARIANT());
+
+	RecursiveUnifiedVectorFormat source_format;
+	Vector::RecursiveToUnifiedFormat(variant, count, source_format);
+
+	//! type_id
+	auto &type_id = UnifiedVariantVector::GetValuesTypeId(source_format);
+	auto type_id_data = type_id.GetData<uint8_t>(type_id);
+
+	//! byte_offset
+	auto &byte_offset = UnifiedVariantVector::GetValuesByteOffset(source_format);
+	auto byte_offset_data = byte_offset.GetData<uint32_t>(byte_offset);
+
+	//! values
+	auto &values = UnifiedVariantVector::GetValues(source_format);
+	auto values_data = values.GetData<list_entry_t>(values);
+
+	//! children
+	auto &children = UnifiedVariantVector::GetChildren(source_format);
+	auto children_data = children.GetData<list_entry_t>(children);
+
+	//! key_id
+	auto &key_id = UnifiedVariantVector::GetChildrenKeyId(source_format);
+	auto key_id_data = key_id.GetData<uint32_t>(key_id);
+
+	//! keys
+	auto &keys = UnifiedVariantVector::GetKeys(source_format);
+	auto keys_data = keys.GetData<list_entry_t>(keys);
+
+	//! keys entry
+	auto &keys_entry = UnifiedVariantVector::GetKeysEntry(source_format);
+	auto keys_entry_data = keys_entry.GetData<string_t>(keys_entry);
+
+	//! value
+	auto &value = UnifiedVariantVector::GetData(source_format);
+	auto value_data = value.GetData<string_t>(value);
+
+	auto result_data = FlatVector::GetData<string_t>(result);
+	for (idx_t i = 0; i < count; i++) {
+		auto index = source_format.unified.sel->get_index(i);
+		if (!source_format.unified.validity.RowIsValid(index)) {
+			result_data[i] = StringVector::AddString(result, "VARIANT_NULL");
```

Comment:
TODO: add this to the internal issue, explain that the root value of a variant can not be VARIANT_NULL, the row's validity instead should be NULL